### PR TITLE
Getting Started guide

### PR DESCRIPTION
**Pull request by Xkonti**

### CODE REVIEW

File: src/content/docs/introduction/getting-started.mdx [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+## What is a val?
+
+A val is a snippet of JavaScript/TypeScript code that can be executed in various ways:
+
+- Manually by clicking the **Run** button
+- Scheduled to run at a specific time or interval
+- Used by other vals
+- Triggered by an email
+- Triggered by an HTTP request (it can even respond to it)
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:25]: This should be one of the first bullets listed.

Also rephrase it to be clearer instead of being cute about responding

File: src/content/docs/introduction/getting-started.mdx [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+In this introduction we will focus on running the val manually and using it in other vals.
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:27]: I think this tutorial covers:

- manually
- imported
- http

File: src/content/docs/introduction/getting-started.mdx [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+- **Run** button - in some cases it might be replaced by a **Save** button.
+- **Add readme** above the **Run** button - allows for adding a short description/documentation to the val.
+
+Under the top section of the val you can see the code editor. For now it contains only a single empty line numbered `1`. This is where we will write our code.
+
+On the bottom of the val there is a status bar showing:
+
+- **Script** - the type of the val. You can change it by clicking on it, but let's leave it at **Script** for now.
+- To the right you can see how many people liked your val (♥️) and how many times it's been referenced by other vals (@).
+- Some other types of vals might also display **Logs** button that will take you to the val's execution logs.
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:61]: This section is already out of date. Try to find a way to write it in a way that won't go out of date very often.

The "Create a val' section may go away so don't mention that. Instead refer to the "New" button in the top right of the screen. That should be more stable.

Don't address the like button or other smaller details like that. Let's keep this focused towards and driving towards a real world example that is live.

File: src/content/docs/introduction/getting-started.mdx [hunk 4]
```diff
@@ -0,0 +1,392 @@
+---
+title: Getting started
+description: |
+  A short guide to get you started with Val Town.
+sidebar:
+  order: 2
+---
+
+Welcome to Val Town! This guide will walk you through the basics of writing vals - snippets of code that are the building blocks of Val Town. We will build a simple web page that allows visitors to subscribe to your newsletter. The website will be composed of multiple vals so that we can demonstrate how they can be used together.
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:9]: I think we should be very thoughtful about what the first thing we want folks to build in VT is:

* website
* cron
* email handler

And to what extent what they create is simply a fork with a couple modifications vs creating it from scratch and copying and pasting from a tutorial.

A component that allows folks to subscribe to an email list is definitely a problem a lot of folks have. However there is a lot of complexity: https://www.petemillspaugh.com/cultivating-emails

A "personal blog" is also something that a lot of people need and is a classic starter project.

More good initial things to do with Val Town ideas: https://www.val.town/v/stevekrouse/welcomeEmail

File: src/content/docs/introduction/getting-started.mdx [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+## Running code manually
+
+It's about time to write some code! Let's start out with the a simple "Hello Val Town" example. Type the following code into the code editor of your val:
+
+```ts val
+console.log("Hello Val Town!");
+```
+
+You should notice that as you're typing, the editor is giving you some hints, just like some other editors you might have used. This may be very helpful, especially when using other vals in your code.
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:71]: Probably can delete this

File: src/content/docs/introduction/getting-started.mdx [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+When you're done typing, click the **Run** button in the top right corner of the val. This will execute the code in the val and display the console output.
+
+![Console output of the "Hello Val Town" val](./getting-started/hello-val-town.png)
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:75]: Delete photos like this that will get stale quickly

File: src/content/docs/introduction/getting-started.mdx [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+
+There it is! Under the code editor you can see the console output of the val saying `Hello Val Town!`. Congratulations! You've just created and executed your first val!
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:77]: Can combine this paragraph with the one above it

File: src/content/docs/introduction/getting-started.mdx [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+
+As you can see the version of the val has changed to **v1**. This is because every time you run a val after making a code change, the val is saved as a new version. This allows for a stress-free experimentation and testing of your code. You can always go back to a previous version of the val if you need to simply by clicking the current version indicator and selecting a different one from a list.
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:79]: We are in the midst of disambiguating running from saving code so try to avoid writing about version numbers too much in this post as it's under flux

Comment (stevekrouse) [reply src/content/docs/introduction/getting-started.mdx:79]: ie just delete this paragraph for now

File: src/content/docs/introduction/getting-started.mdx [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+
+## Building a newsletter subscription page
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:81]: I'm not sure about this section:

1. We should probably use JSX, not an HTML string
2. Introducing JSX and HTML is probably not the first place to go
3. It's boring to make an html/jsx string until you can see it

I think we should start with a http val first and then add html later

File: src/content/docs/introduction/getting-started.mdx [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+Now that we exported a value let's use it in another val.
+
+## Building an API
+
+Besides simply executing code when you click **Run**, Val Town also allows you to create HTTP APIs that are automatically hosted on the web. This means then your code will run whenever someone makes a request to the URL associated with your val. Let's create a val that will return the HTML of our newsletter subscription page.
+
+This time we will create a val from a template. Click the **Templates..** button in the **Create a val** section. This will open up a list of available templates. Scroll fown to **Web apps\* section and click **HTTP server with Hono\*\*. This will create a new val with some code in it:
+
+```ts val
+import { Hono } from "npm:hono@3";
```
Comment (stevekrouse) [inline src/content/docs/introduction/getting-started.mdx:120]: I might avoid hono in the tutorial. Too much random stuff. We can just use an if-statment for routing in this example.

File: src/content/docs/introduction/getting-started.mdx (hunk 4 above)
Comment (stevekrouse) [reply src/content/docs/introduction/getting-started.mdx:9]: Evidence that email collection actually is a good first project: https://twitter.com/TheBuilderJR/status/1740495240304456114?t=xYmfL9c0XzPxCQ3FocUYgg&s=19

File: src/content/docs/introduction/getting-started.mdx (hunk 9 above)
Comment (Xkonti) [reply src/content/docs/introduction/getting-started.mdx:81]: I'd stick to HTML instead of JSX. Everyone can understand HTML string, but not everyone is familiar with JSX (usage/term). I'll think more on how to restructure it to have the order switched.

File: src/content/docs/introduction/getting-started.mdx (hunk 4 above)
Comment (Xkonti) [reply src/content/docs/introduction/getting-started.mdx:9]: I thought a long time on what the "getting started" project should be. My criteria were:
- Displays many strengths of Val Town, such as being super easy and fast to start with, promoting iterative (progressive) workflow, where you can start testing immediately and having various built-in functionalities that don't require prior setup
- Shows how to use other vals that export different types of data, but limit project to just 3 for simplicity
- Teach the most basic built in storage to blow some user's mind that they don't have to configure anything
- Ability to display a website - something REALLY simple that doesn't require anything dynamic but at the same time shows that there is nothing stopping the user from making it dynamic (it's just a string).
- Include some interactivity (user input)
- Ability to respond to requests as most hosting services allow you to host HTML content without a hassle but making an API in an easy way is something less common.

In case of a newsletter signup page I found several pros:
- It's something everyone is familiar with and therefore knows how it works, taking away a cognitive effort
- It's a common problem to solve
- Can be easily expanded into more complex forms, so it's gonna give some ideas
- Can be easily split into 3 tiny vals, where each val is pretty specialized
  - HTTP handler (http val type) - can be complex enough for coding beginners to take a while to understand on it's own
  - HTML content (script val type that exports a value) - you take the ugly strings and lock them away in separate vals - promotes code readability
  - Storage module (script val type that exports a function) - dealing with storage in a separate piece of code to make it easier to understand
- Is open for many easy-to-implement improvements. One of the obvious ones could promote the Pro subscription (senting "thank you for subscribing" emails).
- Can be build in a few steps without a lot of back-and-forth between files.

---

Something like a "personal blog" involves including a lot (more than a few lines) of content that is not Val friendly: HTML or Markdown. I can't see it involve using data storage without advanced querying. Also, can be easily replaced with much easier and more specialized solutions where you just drag your files to upload and have a static website running.

Comment (stevekrouse) [reply src/content/docs/introduction/getting-started.mdx:9]: Just noticing this comment - seems like you can use a lot of this when you close this pr and create a new ticket

**Comment by stevekrouse** [comment]
Closed in favor of: https://github.com/val-town/val-town-docs/issues/87
